src/main/java/uk/gov/justice/digital/delius/service/CustodyService.java

```diff
@@ -108,13 +109,43 @@ public Custody updateCustody(final String nomsNumber,
         });
     }
 
-
+    @Transactional
     public Custody updateCustodyBookingNumber(String nomsNumber, UpdateCustodyBookingNumber updateCustodyBookingNumber) {
-        log.info("Request to update booking number for {}", nomsNumber);
-        return Custody
-                .builder()
-                .bookingNumber(updateCustodyBookingNumber.getBookingNumber())
-                .build();
+        final var maybeOffender = offenderRepository.findByNomsNumber(nomsNumber);
+        final var telemetryProperties = Map.of("offenderNo", nomsNumber,
+                "bookingNumber", updateCustodyBookingNumber.getBookingNumber(),
+                "sentenceStartDate", updateCustodyBookingNumber.getSentenceStartDate().format(DateTimeFormatter.ISO_DATE));
+
+        return maybeOffender.map(offender -> {
+            try {
+                final var maybeEvent = convictionService.getSingleActiveConvictionIdByOffenderIdAndCloseToSentenceDate(offender.getOffenderId(), updateCustodyBookingNumber.getSentenceStartDate());
+                return maybeEvent.map(event -> {
```

Comment:
could do the `orElseThrow` where variable declared - then you wouldn't have to do the `map` and so could avoid large amounts of nesting?

Reply:
Not sure, the problem is essentially we have a few Optionals, but for each one that we done have a value I want to do something slightly different with the Telemetry event.

`maybeA.elseThrow({do something).somethingElse().elseThrow({do something else})` but I will give it a go next revision.

Luckily all these Telemetry events will be removed once I've done and rolled out, so it should become a flat pipeline.

Thanks